lib/Sema/CSGen.cpp

```diff
@@ -2051,11 +2051,30 @@ namespace {
     }
 
     FunctionType *inferClosureType(ClosureExpr *closure) {
-      SmallVector<AnyFunctionType::Param, 4> closureParams;
+      // First let's check whether it could be inferred from the body
+      // that some of the untyped parameters have to be `inout`.
+      if (closure->hasSingleExpressionBody()) {
+        auto *body = closure->getSingleExpressionBody();
+        body->forEachChildExpr([&](Expr *expr) -> Expr * {
+          if (auto *assignment = dyn_cast<AssignExpr>(expr)) {
+            if (auto *DRE = dyn_cast<DeclRefExpr>(assignment->getDest())) {
+              if (auto *PD = dyn_cast_or_null<ParamDecl>(DRE->getDecl())) {
+                if (!PD->getTypeRepr())
+                  PD->setSpecifier(ParamSpecifier::InOut);
```

Comment:
Rather than mutating the `ParamDecl`, should we save the updated specifiers in the constraint system and reference them elsewhere?  We've been moving away from doing mutating within constraint generation.

Reply:
Yeah, this is WIP still, that's what I'm planning to do - record specifiers separately just like we do types, didn't want to make an effort since the benefit might not be there :)

Reply:
Okay, I totally understand! If this is even close to performance-neutral, it's worth it to delete all this code!